### PR TITLE
add katib studyjob launcher

### PR DESCRIPTION
**Pull request by hougangliu**
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/754)

### CODE REVIEW

**Comment by vicaire** [comment]
Hello @hougangliu,

Do you think it would make sense to implement something more generic that could launch and wait for any CRD? 

On top of this generic functionality, some convenience higher level methods could be provided (as the Katib one that you are proposing here). WDYT?

**Comment by hougangliu** [comment]
> Hello @hougangliu,
> 
> Do you think it would make sense to implement something more generic that could launch and wait for any CRD?
> 
> On top of this generic functionality, some convenience higher level methods could be provided (as the Katib one that you are proposing here). WDYT?

It makes sense

**Comment by vicaire** [comment]
/cc @hongye-sun 

Adding Hongye to this thread as he is looking into some related problem.

**Comment by hougangliu** [comment]
@gaoning777 @Ark-kun @hongye-sun any comment?

File: components/kubeflow/katib-launcher/src/launch_study_job.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+
+  expected_condition = ["Completed", "Failed"]
+  wait_response = study_job_client.wait_for_condition(
+      api_client, job_namespace, job_name, expected_condition, 
+      timeout=datetime.timedelta(minutes=args.studyjobtimeoutminutes))
+  succ = False
+  if wait_response.get("status", {}).get("condition") == "Completed":
+    succ = True
+    trial = get_best_trial(wait_response["status"]["bestTrialId"])
+    with open('/output.txt', 'w') as f:
```
Comment (Ark-kun) [inline components/kubeflow/katib-launcher/src/launch_study_job.py]: Please get this path from the command line.

File: components/kubeflow/katib-launcher/src/launch_study_job.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    if value:
+        spec[name] = value
+    else:
+        spec.pop(name)
+
+def _generate_studyjob_yaml(src_filename, name, namespace, optimizationtype, objectivevaluename, optimizationgoal, requestcount,
+                            metricsnames, parameterconfigs, nasConfig, workertemplatepath, mcollectortemplatepath, suggestionspec):
+  """_generate_studyjob_yaml generates studyjob yaml file based on hp.template.yaml"""
+  with open(src_filename, 'r') as f:
+    content = yaml.load(f)
```
Comment (Ark-kun) [inline components/kubeflow/katib-launcher/src/launch_study_job.py]: Maybe you can just fully build the spec dict here in code instead of loading it from file and replacing some values?

Comment (hougangliu) [reply components/kubeflow/katib-launcher/src/launch_study_job.py]: just follow how tfjob-launcher fills spec dict. I will submit another PR to fix the both two components. Thanks!

File: components/kubeflow/katib-launcher/src/launch_study_job.py (hunk 1 above)
Comment (hougangliu) [reply components/kubeflow/katib-launcher/src/launch_study_job.py]: done! thanks!

File: .cloudbuild.yaml [hunk 3]
```diff
@@ -97,8 +97,12 @@ steps:
   id: 'buildDeployer'
 - name: 'gcr.io/cloud-builders/docker'
   entrypoint: '/bin/bash'
-  args: ['-c', 'cd /workspace/components/kubeflow/launcher && ./build_image.sh -p $PROJECT_ID -t $COMMIT_SHA']
```
Comment (vicaire) [inline .cloudbuild.yaml]: Could we not build the container automatically for now? It requires a complicated review process on our side.

File: .cloudbuild.yaml [hunk 4]
```diff
@@ -97,8 +97,12 @@ steps:
...
+  args: ['-c', 'cd /workspace/components/kubeflow/tf-launcher && ./build_image.sh -p $PROJECT_ID -t $COMMIT_SHA']
```
Comment (vicaire) [inline .cloudbuild.yaml]: Is the change from launcher to tf-launcher necessary? It seems unrelated to this PR.

**Comment by vicaire** [comment]
@hongye-sun, could you please have a look at this one? I think you could have some good feedback as you are working on many components.

File: .cloudbuild.yaml (hunk 3 above)
Comment (hougangliu) [reply .cloudbuild.yaml]: I just rename launcher (in fact, it only launches tfjob) to tf-launcher since we can add more launchers (saying pytorch job launcher, studyjob launcher and so on)

File: .cloudbuild.yaml (hunk 4 above)
Comment (hougangliu) [reply .cloudbuild.yaml]: I just rename launcher (in fact, it only launches tfjob) to tf-launcher since we can add more launchers (saying pytorch job launcher, studyjob launcher and so on)

File: .cloudbuild.yaml (hunk 3 above)
Comment (vicaire) [reply .cloudbuild.yaml]: Thank you @hougangliu. Could we do the renaming in a separate PR? There are several dependencies on this tfjob component so I would like to separate the renaming into another PR in case we need to rollback.

File: .cloudbuild.yaml [hunk 5]
```diff
@@ -97,8 +97,12 @@ steps:
...
   id: 'buildLauncher'
+- name: 'gcr.io/cloud-builders/docker'
+  entrypoint: '/bin/bash'
```
Comment (vicaire) [inline .cloudbuild.yaml]: Hi @hougangliu,

This entry is building the new container automatically. Could you please remove this entry? It requires a complicated review process on our side before we can release new containers automatically.

Comment (hougangliu) [reply .cloudbuild.yaml]: thanks, Done
